src/include/duckdb/common/cycle_counter.hpp

```diff
@@ -10,32 +10,72 @@
 
 #include "duckdb/common/helper.hpp"
 #include "duckdb/common/chrono.hpp"
+#include "duckdb/common/random_engine.hpp"
 
 namespace duckdb {
 
-//! The cycle counter can be used to measure elapsed cycles
+//! The cycle counter can be used to measure elapsed cycles for a function, expression and ...
+//! Optimized by sampling mechanism. Once per 100 times.
+//! //Todo Can be optimized further by calling RDTSC once per sample
 class CycleCounter {
 public:
-	//! Starts the timer
-	void Start() {
-		finished = false;
-		start = Tick();
+	CycleCounter() : random(0) {
 	}
-	//! Finishes timing
-	void End() {
-		end = Tick();
-		finished = true;
+	// Next_sample determines if a sample needs to be taken, if so start the profiler
+	void TakeSample() {
+		if (current_count >= next_sample) {
+			tmp = Tick();
+		}
 	}
 
-	uint64_t Elapsed() const {
-		return end - start;
+	// Process the sample
+	void ProcessSample(int chunk_size) {
+		if (current_count >= next_sample) {
```

Comment:
This code block seems like a repeat of ProcessSample(). Can this function just call ProcessSample()?

Reply:
This is like an optimization. because for a function inside an expression, we don't need to count the number of process tuples. it would happen already in the expression-root cycle counter.

Reply:
That's fine, but that doesn't require the code to be copy pasted, no?

Also care would need to be taken in that case, since tuple_count will be zero on the expression counter, right? There are several places where we divide by the tuple_count, which could lead to a division by zero.

Reply:
The tuple_count will be zero, but it will be never used. The tuple_count from Expressionexecutorstate (expression_root) will be used in those cases. Potentially the function cycle-counter and root-expression cycle-counter could be divided into two classes which one inherit from the other one. This way, we also save space.